app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.cpp

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include "IHardwareOpusDecoder.h"
+#include "results.h"
+#include <services/timesrv/common.h>
+
+namespace skyline::service::codec {
+    size_t CalculateOutBufferSize(i32 sampleRate, i32 channelCount, i32 frameSize) {
+        return util::AlignUp(frameSize * channelCount / (OpusFullbandSampleRate / sampleRate), 0x40);
+    }
+
+    IHardwareOpusDecoder::IHardwareOpusDecoder(const DeviceState &state, ServiceManager &manager, i32 sampleRate, i32 channelCount) : BaseService(state, manager), sampleRate(sampleRate), channelCount(channelCount), decoderOutputBufferSize(CalculateOutBufferSize(sampleRate, channelCount, MaxFrameSizeNormal)) {
+        int result;
+        decoderState = opus_decoder_create(sampleRate, channelCount, &result);
+        if (result != OPUS_OK)
+            throw OpusException(result);
+    }
+
+    IHardwareOpusDecoder::~IHardwareOpusDecoder() {
+        opus_decoder_destroy(decoderState);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleavedImpl(request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleavedImpl(request, response, true);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfAndResetOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleaved(session, request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleaved(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto reset{request.Pop<bool>()};
+        if (reset)
+            ResetContext();
+
+        return DecodeInterleavedImpl(request, response, true);
+    }
+
+    void IHardwareOpusDecoder::ResetContext() {
+        opus_decoder_ctl(decoderState, OPUS_RESET_STATE);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedImpl(ipc::IpcRequest &request, ipc::IpcResponse &response, bool performanceInfo) {
+        auto dataIn{request.inputBuf.at(0)};
+        auto dataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        if (dataIn.size() <= sizeof(OpusDataHeader))
+            return result::BadOpusDataHeader;
```

Comment:
This should throw an exception and so should the other one making this function return void, alternatively make this function return a result and then have all things that call it throw an exception on Nonzero value